pkg/apis/app/v1beta1/application_types.go

```diff
@@ -84,15 +84,34 @@ type ApplicationStatus struct {
 
 // ImageSpec contains information about an image used as an icon.
 type ImageSpec struct {
-	// The source for image represented as either an absolute URL to the image or a Data URL containing
+	// Source for the image represented as either an absolute URL to the image or a Data URL containing
 	// the image. Data URLs are defined in RFC 2397.
-	Source string `json:"src"`
+	//
+	// Only one of Source or Data may be defined.
+	Source string `json:"src,omitempty"`
 
-	// (optional) The size of the image in pixels (e.g., 25x25).
+	// Size of the image in pixels (e.g., 25x25). Optional.
 	Size string `json:"size,omitempty"`
 
-	// (optional) The mine type of the image (e.g., "image/png").
+	// Type indicates the MIME type of the image (e.g., "image/png"). Optinal.
 	Type string `json:"type,omitempty"`
+
+	// Data contains the binary information for an image.
+	//
+	// Only one of Source or Data may be defined.
+	Data *ImageData `json:"data,omitempty"`
+}
+
+// ImageData contains raw information for the image, either as binary, text, or a reference to a ConfigMap entry.
+type ImageData struct {
+	// Binary contains the base64-encoded image binary.
+	Binary string `json:"binary,omitempty"`
+
+	// Text contains the image definition as plain text, for formats such as SVG.
+	Text string `json:"text,omitempty"`
+
+	// ConfigMapKeyRef is a reference to a ConfigMap entry containing the image data.
+	ConfigMapKeyRef *ConfigMapKeySelector `json:"configMapKeyRef,omitempty"`
```

Comment:
why not just move the "source url" in here too?  You still only want exactly one of these to be filled in.

```
type Image struct {
  Binary string
  Text string
  URL string
  ConfigMapKeyRef *ConfigMapKeySelector
}
```

Reply:
Certainly doable. I was going for a syntax along the lines of EnvVar in the ContainerSpec. Curious if this syntax remains in the v2 API work.

I'd probably keep `Source`, though, as this has merged and therefore "landed" as the API.

Reply:
it's a beta, i don't think it having landed is a good reason to feel like we're stuck with it if there is agreement on a better approach before it becomes finalized.